File: 자바스크립트/캐시.js
function solution(cacheSize, cities) {
    let cache = [];
    let answer = 0
    const lowerCities = cities.map(city => city.toLowerCase())

    if (cacheSize === 0) return cities.length * 5
    for (const city of lowerCities) {
        // 캐시위에 있을 때 (hit)
        if (cache.indexOf(city) !== -1) {
            cache = cache.filter(c => c !== city).concat(city)
            answer += 1
            continue;
        }
        // (miss)
        if (cache.length === cacheSize) cache.shift()
        cache.push(city)
        answer += 5
    }
    return answer;
}

File: 캐시.py
from collections import deque


def solution(cacheSize, cities):
    cache = deque([])
    answer = 0

    if cacheSize == 0:  # 캐시가 없을 때
        return len(cities) * 5

    for city in cities:
        city = city.lower()  # 대소문자 구분없음
        # 캐시위에 있을 때 (hit)
        if city in cache:
            cache.remove(city)
            cache.append(city)
            answer += 1
            continue
        # (miss)
        if len(cache) == cacheSize:  # 자리가 없다면
            cache.popleft()
        cache.append(city)
        answer += 5
    return answer


'''
import heapq
def solution(cacheSize, cities): # 11, 15, 18, 19 실패
    TIME, CITY = 0, 1
    cache = []  # 시간과 함께 저장
    cur_cache = set()  # 캐시위에 올라와있는 도시
    answer = 0

    if cacheSize == 0:  # 캐시가 없을 때
        return len(cities) * 5

    for time, city in enumerate(cities, 1):
        city = city.lower()

        # 캐시위에 있을 때 (hit)
        if city in cur_cache:
            for i, c in enumerate(cache):
                if c[CITY] == city:
                    cache[i] = (time, city)
                    break
            answer += 1
            continue

        # (miss)
        if len(cache) == cacheSize:  # 자리가 없다면
            cur_cache.remove(heapq.heappop(cache)[CITY])
        heapq.heappush(cache, (time, city))
        cur_cache.add(city)
        answer += 5

    return answer
'''
